File: edge/worker/src/challenge.ts
/**
 * fox-shield edge worker — progressive proof-of-work challenge.
 *
 * Serves /__shield/challenge. GET returns an HTML page whose JavaScript finds a
 * nonce such that SHA-256(nonce + ip) starts with `difficulty` leading zero
 * hex characters (4 normal / 5 aggressive). The client POSTs the nonce; on
 * verification a __shield_pass cookie is set (10m expiry) and the request is
 * allowed through. If the proof is invalid the client is served a CAPTCHA
 * placeholder (a simple math challenge) as a fallback.
 *
 * Mirrors internal/challenge/challenge.go.
 */

import type { Store } from './store';

export interface ChallengeConfig {
  difficultyNormal: number;
  difficultyAggressive: number;
  passTtlSeconds: number;
}

export const DEFAULT_CHALLENGE_CONFIG: ChallengeConfig = {
  difficultyNormal: 4,
  difficultyAggressive: 5,
  passTtlSeconds: 10 * 60,
};

export const CHALLENGE_COOKIE = '__shield_pass';

/** Returns a random hex string of `bytes` bytes. */
export function randomHex(bytes: number): string {
  const arr = new Uint8Array(bytes);
  crypto.getRandomValues(arr);
  return [...arr].map((b) => b.toString(16).padStart(2, '0')).join('');
}

/** SHA-256 hex digest of the input. */
export async function sha256Hex(input: string): Promise<string> {
  const data = new TextEncoder().encode(input);
  const digest = await crypto.subtle.digest('SHA-256', data);
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

/** Verifies that sha256(nonce + proof) starts with `difficulty` zero hex chars. */
export async function validProof(nonce: string, proof: string, difficulty: number): Promise<boolean> {
  if (nonce === '' || proof === '') {
    return false;
  }
  const hex = await sha256Hex(nonce + proof);
  return hex.startsWith('0'.repeat(difficulty));
}

/** Builds the challenge HTML page with embedded proof-of-work JavaScript. */
export function challengeHtml(nonce: string, difficulty: number, ip: string): string {
  const script = `
    const nonce = ${JSON.stringify(nonce)};
    const difficulty = ${difficulty};
    const ip = ${JSON.stringify(ip)};
    async function sha256hex(s) {
      const buf = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(s));
      return [...new Uint8Array(buf)].map(b => b.toString(16).padStart(2, '0')).join('');
    }
    async function solve() {
      const target = '0'.repeat(difficulty);
      const status = document.getElementById('status');
      for (let proof = 0; proof < 100000000; proof++) {
        const hex = await sha256hex(nonce + proof);
        if (hex.startsWith(target)) {
          const res = await fetch('/__shield/challenge', {
            method: 'POST',
            headers: { 'content-type': 'application/json' },
            body: JSON.stringify({ nonce, proof: String(proof) }),
          });
          if (res.ok) { location.reload(); }
          else { status.textContent = 'Proof rejected — retrying.'; }
          return;
        }
        if (proof % 10000 === 0) {
          status.textContent = 'Solving proof-of-work… ' + proof;
          await new Promise(r => setTimeout(r, 0));
        }
      }
      status.textContent = 'Could not solve — please refresh.';
    }
    solve();
  `;
  return `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>fox-shield — verifying you are human</title>
<style>
  body { font-family: system-ui, sans-serif; background: #0f1115; color: #e6e8eb;
         display: flex; align-items: center; justify-content: center; min-height: 100vh; margin: 0; }
  .card { background: #1a1d24; border: 1px solid #2a2f3a; border-radius: 12px;
          padding: 32px; max-width: 420px; text-align: center; }
  h1 { font-size: 20px; margin: 0 0 8px; }
  p { color: #9aa3b2; font-size: 14px; line-height: 1.5; }
  #status { color: #7fb3ff; font-size: 13px; margin-top: 16px; min-height: 18px; }
  .spinner { width: 28px; height: 28px; border: 3px solid #2a2f3a; border-top-color: #7fb3ff;
             border-radius: 50%; margin: 16px auto 0; animation: spin 0.8s linear infinite; }
  @keyframes spin { to { transform: rotate(360deg); } }
</style>
</head>
<body>
  <div class="card">
    <h1>Verifying you are human</h1>
    <p>fox-shield is checking your browser with a proof-of-work challenge. This should complete in a few seconds.</p>
    <div class="spinner"></div>
    <div id="status">Starting…</div>
  </div>
  <script>${script}</script>
</body>
</html>`;
}

/** Builds the CAPTCHA placeholder page (simple math challenge). */
export function captchaHtml(ip: string): string {
  const a = Math.floor(Math.random() * 9) + 1;
  const b = Math.floor(Math.random() * 9) + 1;
  const answer = a + b;
  return `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>fox-shield — CAPTCHA</title>
<style>
  body { font-family: system-ui, sans-serif; background: #0f1115; color: #e6e8eb;
         display: flex; align-items: center; justify-content: center; min-height: 100vh; margin: 0; }
  .card { background: #1a1d24; border: 1px solid #2a2f3a; border-radius: 12px; padding: 32px; max-width: 360px; }
  h1 { font-size: 20px; margin: 0 0 16px; }
  form { display: flex; flex-direction: column; gap: 12px; }
  input { padding: 10px; border-radius: 8px; border: 1px solid #2a2f3a; background: #0f1115; color: #e6e8eb; }
  button { padding: 10px; border-radius: 8px; border: 0; background: #7fb3ff; color: #0f1115; font-weight: 600; cursor: pointer; }
</style>
</head>
<body>
  <div class="card">
    <h1>Human verification</h1>
    <p>Please solve: ${a} + ${b} = ?</p>
    <form method="POST" action="/__shield/challenge">
      <input type="hidden" name="captcha" value="${answer}">
      <input type="hidden" name="ip" value="${ip}">
      <input type="number" name="answer" required autofocus placeholder="Your answer">
      <button type="submit">Verify</button>
    </form>
  </div>
</body>
</html>`;
}

export class Challenge {
  private readonly cfg: ChallengeConfig;

  constructor(cfg: ChallengeConfig = DEFAULT_CHALLENGE_CONFIG) {
    this.cfg = cfg;
  }

  private difficulty(aggressive: boolean): number {
    return aggressive ? this.cfg.difficultyAggressive : this.cfg.difficultyNormal;
  }

  /** Serves the challenge page for a GET request. */
  serve(ip: string, aggressive: boolean): Response {
    const nonce = randomHex(16);
    const html = challengeHtml(nonce, this.difficulty(aggressive), ip);
    return new Response(html, {
      status: 200,
      headers: { 'content-type': 'text/html; charset=utf-8', 'cache-control': 'no-store' },
    });
  }

  /**
   * Verifies a submitted proof (JSON { nonce, proof } or CAPTCHA form data).
   * On success sets the __shield_pass cookie and returns a 200; otherwise
   * returns the CAPTCHA placeholder.
   */
  async verify(request: Request, ip: string, aggressive: boolean): Promise<Response> {
    const contentType = request.headers.get('content-type') ?? '';

    if (contentType.includes('application/json')) {
      let nonce = '';
      let proof = '';
      try {
        const body: unknown = await request.json();
        if (typeof body === 'object' && body !== null) {
          const rec = body as Record<string, unknown>;
          if (typeof rec.nonce === 'string') nonce = rec.nonce;
          if (typeof rec.proof === 'string') proof = rec.proof;
        }
      } catch {
        return new Response('Bad Request', { status: 400 });
      }
      if (await validProof(nonce, proof, this.difficulty(aggressive))) {
        return this.passResponse();
      }
      // Proof failed -> CAPTCHA placeholder fallback.
      return new Response(captchaHtml(ip), {
        status: 403,
        headers: { 'content-type': 'text/html; charset=utf-8', 'cache-control': 'no-store' },
      });
    }

    // Form-encoded CAPTCHA submission.
    const form = await request.formData();
    const expected = form.get('captcha');
    const answer = form.get('answer');
    if (typeof expected === 'string' && typeof answer === 'string' && expected === answer) {
      return this.passResponse();
    }
    return new Response(captchaHtml(ip), {
      status: 403,
      headers: { 'content-type': 'text/html; charset=utf-8', 'cache-control': 'no-store' },
    });
  }

  private passResponse(): Response {
    const cookie = `${CHALLENGE_COOKIE}=1; Path=/; Max-Age=${this.cfg.passTtlSeconds}; HttpOnly; SameSite=Lax`;
    return new Response('ok', {
      status: 200,
      headers: { 'set-cookie': cookie, 'cache-control': 'no-store' },
    });
  }
}

File: internal/challenge/challenge.go
// Package challenge implements a JavaScript proof-of-work challenge. The
// shield serves a challenge page at /__shield/challenge that generates a
// random nonce; the client must find a value whose SHA-256 hash starts with a
// configurable number of leading zero bits. On success a __shield_pass cookie
// is set and the client is allowed through.
package challenge

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"strings"
	"time"

	"github.com/foxai/fox-shield/internal/limiter"
	"github.com/foxai/fox-shield/internal/store"
)

// Config holds challenge parameters.
type Config struct {
	// DifficultyNormal is the number of leading zero bits required in normal
	// mode.
	DifficultyNormal int
	// DifficultyAggressive is the difficulty in aggressive mode.
	DifficultyAggressive int
	// PassTTL is how long a solved __shield_pass cookie is valid.
	PassTTL time.Duration
}

// DefaultConfig returns the v1.0 defaults.
func DefaultConfig() Config {
	return Config{
		DifficultyNormal:     4,
		DifficultyAggressive: 5,
		PassTTL:              10 * time.Minute,
	}
}

// CookieName is the name of the proof-of-work pass cookie.
const CookieName = "__shield_pass"

// Challenge issues and verifies proof-of-work challenges.
type Challenge struct {
	cfg   Config
	store store.Store
}

// New creates a Challenge.
func New(cfg Config, s store.Store) *Challenge {
	return &Challenge{cfg: cfg, store: s}
}

// Handler serves the challenge endpoint. It returns a JSON payload with the
// nonce and difficulty, and verifies submitted solutions.
func (c *Challenge) Handler() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodPost {
			c.verify(w, r)
			return
		}
		nonce := randomHex(16)
		difficulty := c.cfg.DifficultyNormal
		if r.Context().Value(ctxKeyAggressive) == true {
			difficulty = c.cfg.DifficultyAggressive
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]interface{}{
			"nonce":      nonce,
			"difficulty": difficulty,
			"algorithm":  "sha256",
		})
	})
}

func (c *Challenge) verify(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Nonce string `json:"nonce"`
		Proof string `json:"proof"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Bad Request", http.StatusBadRequest)
		return
	}
	difficulty := c.cfg.DifficultyNormal
	if r.Context().Value(ctxKeyAggressive) == true {
		difficulty = c.cfg.DifficultyAggressive
	}
	if !validProof(req.Nonce, req.Proof, difficulty) {
		http.Error(w, "Invalid Proof", http.StatusForbidden)
		return
	}
	http.SetCookie(w, &http.Cookie{
		Name:     CookieName,
		Value:    randomHex(16),
		Path:     "/",
		MaxAge:   int(c.cfg.PassTTL.Seconds()),
		HttpOnly: true,
		SameSite: http.SameSiteLaxMode,
	})
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write([]byte("ok"))
}

// validProof checks that sha256(nonce+proof) has `difficulty` leading zero bits.
func validProof(nonce, proof string, difficulty int) bool {
	if nonce == "" || proof == "" {
		return false
	}
	sum := sha256.Sum256([]byte(nonce + proof))
	hexStr := hex.EncodeToString(sum[:])
	// Count leading zero bits from the hex string.
	bits := 0
	for i := 0; i < len(hexStr); i++ {
		v := hexVal(hexStr[i])
		if v == 0 {
			bits += 4
			continue
		}
		// Count leading zero bits of this nibble.
		for shift := 3; shift >= 0; shift-- {
			if v&(1<<shift) == 0 {
				bits++
			} else {
				break
			}
		}
		break
	}
	return bits >= difficulty
}

func hexVal(c byte) int {
	switch {
	case c >= '0' && c <= '9':
		return int(c - '0')
	case c >= 'a' && c <= 'f':
		return int(c-'a') + 10
	case c >= 'A' && c <= 'F':
		return int(c-'A') + 10
	}
	return 0
}

// Middleware requires a valid pass cookie for suspicious requests, otherwise
// it serves the challenge. Requests not marked suspicious pass through.
func (c *Challenge) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if strings.HasPrefix(r.URL.Path, "/__shield/") {
			next.ServeHTTP(w, r)
			return
		}
		if !limiter.IsSuspicious(r.Context()) {
			next.ServeHTTP(w, r)
			return
		}
		if cookie, err := r.Cookie(CookieName); err == nil && cookie.Value != "" {
			next.ServeHTTP(w, r)
			return
		}
		c.Handler().ServeHTTP(w, r)
	})
}

func randomHex(n int) string {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "0000000000000000"
	}
	return hex.EncodeToString(b)
}

type ctxKey int

const ctxKeyAggressive ctxKey = 0

// WithAggressive returns a context carrying the aggressive-mode flag.
func WithAggressive(ctx context.Context, aggressive bool) context.Context {
	return context.WithValue(ctx, ctxKeyAggressive, aggressive)
}
